opentelemetry-etw-metrics/src/exporter/mod.rs

```diff
@@ -0,0 +1,90 @@
+use opentelemetry::metrics::{MetricsError, Result};
+use opentelemetry_proto::tonic::collector::metrics::v1::ExportMetricsServiceRequest;
+use opentelemetry_sdk::metrics::{
+    data::{ResourceMetrics, Temporality},
+    exporter::PushMetricsExporter,
+    reader::{AggregationSelector, DefaultAggregationSelector, TemporalitySelector},
+    Aggregation, InstrumentKind,
+};
+
+use async_trait::async_trait;
+use prost::Message;
+use tracelogging_dynamic as tld;
+
+use std::fmt::{Debug, Formatter};
+
+use crate::etw;
+
+pub struct MetricsExporter {
+    provider: etw::ProviderGuard,
+}
+
+impl MetricsExporter {
+    pub fn new(
+        name: &str,
+        id: Option<tld::Guid>,
+        options: Option<tld::ProviderOptions>,
+    ) -> Result<MetricsExporter> {
+        Ok(MetricsExporter {
+            provider: etw::ProviderGuard::register(name, id, options)?,
+        })
+    }
+}
+
+impl TemporalitySelector for MetricsExporter {
+    fn temporality(&self, kind: InstrumentKind) -> Temporality {
+        match kind {
+            InstrumentKind::Counter
+            | InstrumentKind::ObservableCounter
+            | InstrumentKind::ObservableGauge
+            | InstrumentKind::Histogram
+            | InstrumentKind::Gauge => Temporality::Delta,
+            InstrumentKind::UpDownCounter | InstrumentKind::ObservableUpDownCounter => {
+                Temporality::Cumulative
+            }
+        }
+    }
+}
+
+impl AggregationSelector for MetricsExporter {
+    fn aggregation(&self, kind: InstrumentKind) -> Aggregation {
+        DefaultAggregationSelector::new().aggregation(kind)
+    }
+}
+
+impl Debug for MetricsExporter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str("ETW metrics exporter")
+    }
+}
+
+#[async_trait]
+impl PushMetricsExporter for MetricsExporter {
+    async fn export(&self, metrics: &mut ResourceMetrics) -> Result<()> {
+        let proto_message: ExportMetricsServiceRequest = (&*metrics).into();
+
+        let mut byte_array = Vec::new();
+        proto_message
+            .encode(&mut byte_array)
+            .map_err(|err| MetricsError::Other(err.to_string()))?;
+
+        let result = self.provider.write(&byte_array);
+
+        match result {
+            0 => println!("Successfully wrote ETW event"),
```

Comment:
lets use the built-in error handling.

Reply:
I updated the error path to set the global error on `unregister` failure